cilium/CFP-22215-mutual-auth-for-service-mesh.md

```diff
@@ -0,0 +1,455 @@
+* When the datapath generates a datapath drop message with the reason that "authentication is required" for this peer, look up the corresponding node and establish an mTLS connection to the peer node. Connects to the port outlined in the Startup section above.
+  * Use the Cilium Identity to look up TLS details.
+  * Special case: peer node may be the same node as the current node. In this case, the Node can connect to its own serving port, and perform the handshake, just like a remote node.
+  * Validate certificate of the peer. Note that since we are also planning to handle re-key events by pre-emptively re-authenticating, this should not occur very often.
+    * Q: How to make invalid certificate errors super obvious for debugging?
+      * Logs
+      * Metrics
+
+##### Authenticated session established with peer
+
+* Update datapath to inform it that this peer is authenticated (see Data Plane for details)
+* The destination node is responsible for checking that a Pod matching the destination Identity is running on itself, and only allowing the authentication to succeed if it is.
+
+##### Endpoint deletion
+
+* Node with endpoint must clear out Auth table to remove any state that implies authentication occurred for connections with this endpoint (This should already exist)
+* Do all other nodes need to clear out their corresponding state also?
+  * Thinking of IP recycling case, e.g. in ENI mode an IP could move from one node to another.
+  * Trigger on CiliumEndpoint deletion even for pods on remote nodes, delete authentication state corresponding to that peer. Including userspace authentication table entries, datapath CT/map entries.
+
+##### Timer-based
+* Do we need periodic garbage collection, e.g. of authenticated sessions? Because the auth table has an expiry time, we have built-in garbage collection - the userspace will be responsible for pruning expired auth table entries if the identity has been deleted and the certificate validity period has passed.
+* At this stage, it seems likely that deletion of expired entries will still be done by userspace, although this could change in the future.
+
+#### Certificate management
+
+Initial certificate management for this feature will be using SPIRE and SPIFFE, with interfaces built to allow us to walk back to other options once we have more user intent data.
+
+### Data Plane
+
+
+#### Structures
+Policy Entry - 1 bit: "Authentication Required"
+Drop events - 1 codepoint - "Authentication Required"
+Dedicated bpf auth map, see Key question: How to implement dataplane auth check
+
+#### Logic
+When performing policy lookup, check 'authentication required' bit.
+If this bit is set and there is no corresponding entry in the BPF auth table, drop the traffic and emit an 'authentication required' event to the local control plane.
+Ensure both egress and ingress implement this logic.
+For source and destination on the same node, the node will connect to itself and perform a handshake.
+
```

Comment:
A couple of cases are missing here:
- When auth check passes (simple - allow the packet through.)
- Repeat the process on ingress (is it a direct repeat of the process on egress?)

Reply:
I guess these are covered in the lifecycle / operations walkthrough anyway so no need to include here.